R/derive_extreme_records.R

```diff
@@ -101,42 +177,165 @@
+                                   dataset_ref = NULL,
                                    by_vars = NULL,
-                                   order,
-                                   mode,
+                                   order = NULL,
+                                   mode = NULL,
+                                   filter_add = NULL,
                                    check_type = "warning",
-                                   filter = NULL,
-                                   set_values_to) {
-  # Check input parameters
-  assert_vars(by_vars, optional = TRUE)
-  assert_order_vars(order)
+                                   exist_flag = NULL,
+                                   true_value = "Y",
+                                   false_value = "N",
+                                   set_values_to,
+                                   filter) {
+  if (!missing(filter)) {
+    deprecate_warn(
+      "0.11.0",
+      "derive_extreme_records(filter = )",
+      "derive_extreme_records(filter_add = )"
+    )
+    filter_add <- enexpr(filter)
+  }
+
+  # Check input arguments
+  assert_vars(by_vars, optional = is.null(dataset_ref))
+  assert_order_vars(order, optional = TRUE)
   assert_data_frame(
     dataset,
     required_vars = expr_c(
       by_vars, extract_vars(order)
-    )
+    ),
+    optional = TRUE
+  )
+  mode <- assert_character_scalar(
+    mode,
+    values = c("first", "last"),
+    case_sensitive = FALSE,
+    optional = TRUE
   )
-  mode <- assert_character_scalar(mode, values = c("first", "last"), case_sensitive = FALSE)
   check_type <-
     assert_character_scalar(
       check_type,
       values = c("none", "warning", "error"),
       case_sensitive = FALSE
     )
-  filter <- assert_filter_cond(enexpr(filter), optional = TRUE)
+  exist_flag <- assert_symbol(enexpr(exist_flag), optional = TRUE)
+  filter_add <- assert_filter_cond(enexpr(filter_add), optional = TRUE)
   assert_varval_list(set_values_to)
+  if (is.null(dataset) && is.null(dataset_add)) {
+    abort(paste(
+      "Neither `dataset` nor `dataset_add` is specified.",
+      "At least one of them must be specified.",
+      sep = "\n"
+    ))
+  }
 
   # Create new observations
-  new_obs <- dataset %>%
-    filter_if(filter) %>%
-    filter_extreme(
+  if (is.null(dataset_add)) {
+    dataset_add <- dataset
+  }
+  new_add_obs <- filter_if(dataset_add, filter_add)
+
+  if (!is.null(order)) {
+    new_add_obs <- filter_extreme(
+      new_add_obs,
       by_vars = by_vars,
       order = order,
       mode = mode,
       check_type = check_type
-    ) %>%
-    mutate(!!!set_values_to)
+    )
+  }
+
+  if (!is.null(dataset_ref)) {
+    add_vars <- colnames(dataset_add)
+    ref_vars <- colnames(dataset_ref)
+
+    new_ref_obs <- anti_join(
+      select(dataset_ref, intersect(add_vars, ref_vars)),
+      select(new_add_obs, !!!by_vars),
+      by = sapply(by_vars, as_name) # nolint: undesirable_function_linter
```

Comment:
Is this `nolint` still needed?

Reply:
I have replaced `sapply()` with `map_chr()`. Then we do not need `nolint`.